### PR TITLE
improve worst-case performance of BTreeSet intersection v1

### PR DESCRIPTION
**Pull request by ssomers**
Major performance boost when comparing tiny and huge sets. Probably with controversial changes and I sure have questions:

- Names and places of functions and types
- How many comments to write where
- Why does rustc tell me to `ref mut` and `ref` matches on the iterator, while the book says ref is old school: fixed in code, should take it up to compiler people.
- (Why) do I have to write out the clone like that (`#[derive(Clone)]` doesn't work): see https://github.com/rust-lang/rust/pull/59186#issuecomment-477942422_
- Am I allowed to use `#[derive(Debug)]` there at all? Yes, but the actual Debug.fmt implementation should supersede it.
- I'd like to test function `are_proportionate_for_intersection` in test_intersection (or another test case next to it) itself, but I think the private function is inaccessible there. liballoc has other test cases not in the tests directory.

PS I don't list these questions to start a discussion here, just to inspire reviewers and to remember myself.

### CODE REVIEW

File: src/liballoc/collections/btree/set.rs [hunk 1]
```diff
@@ -155,6 +155,29 @@ impl<T: fmt::Debug> fmt::Debug for SymmetricDifference<'_, T> {
     }
 }
 
+#[stable(feature = "rust1", since = "1.0.0")]
```
Comment (Centril) [inline src/liballoc/collections/btree/set.rs]: Remove this line; this is not a public API you're exposing.

File: src/liballoc/collections/btree/set.rs [hunk 2]
```diff
@@ -1069,6 +1103,15 @@ impl<'a, T: Ord> Iterator for SymmetricDifference<'a, T> {
 #[stable(feature = "fused", since = "1.26.0")]
 impl<T: Ord> FusedIterator for SymmetricDifference<'_, T> {}
 
+#[stable(feature = "rust1", since = "1.0.0")]
```
Comment (Centril) [inline src/liballoc/collections/btree/set.rs]: and here too

File: src/liballoc/collections/btree/set.rs [hunk 3]
```diff
@@ -155,6 +155,28 @@ impl<T: fmt::Debug> fmt::Debug for SymmetricDifference<'_, T> {
     }
 }
 
+#[derive(Debug)]
+enum IntersectionOther<'a, T> {
+    ITER(Peekable<Iter<'a, T>>),
```
Comment (varkor) [inline src/liballoc/collections/btree/set.rs]: Enum variants should be `UpperCamelCase`.

**Comment by KodrAus** [comment]
Thanks @ssomers!

Would you be able to put together some benchmarks so we could get an idea of what the impact is for intersecting variously sized sets?

**Comment by ssomers** [comment]
> put together some benchmarks so we could get an idea of what the impact is for intersecting variously sized sets?

I'm not sure how to contrast the various implementations in the same build. But I could first PR a benchmark in src\liballoc\benches\btree using the existing implementation, then merge this here and anyone looking closely enough would see times improve.

**Comment by ssomers** [comment]
Rough numbers on my system:

bset::tree ttest case|ns/iter before this PR|ns/iter after this PR
--|--|--     
intersect_neg_vs_pos_100|500|=
intersect_neg_vs_pos_10_vs_10k|60|150
intersect_neg_vs_pos_10k|50,000|=
intersect_neg_vs_pos_10k_vs_10|50,000|400
intersect_pos_vs_neg_100|400|=
intersect_pos_vs_neg_10_vs_10k|40,000|150
intersect_pos_vs_neg_10k|40,000|=
intersect_pos_vs_neg_10k_vs_10|60|400
intersect_random_100|700|=
intersect_random_10_vs_10k|40,000|300
intersect_random_10k|200,000|=
intersect_random_10k_vs_10|50,000|300
intersect_staggered_100|700|=
intersect_staggered_10_vs_10k|80|150
intersect_staggered_10k|70,000|=
intersect_staggered_10k_vs_10|80|150

As you see, intersection of very asymmetric set (10 versus 10,000 elements) is mostly up, except if the set containing only numbers lower than any in the other set (neg = negative ints vs. pos=positive ints) is also a small set (i.e. the algorithm merely iterates over that small set and finishes).

The latter could be improved upon by limiting both sets to the range started with the smallest element of the other set. The whole algorithm could work like that: instead of iterating, find the next element >= the other set's lowest unseen value. But then sometimes/often the binary search towards the next element would be worse than iterating, in particular for the staggered test case.

**Comment by KodrAus** [comment]
Thanks for putting the numbers together @ssomers. That's much appreciated.

It looks like we aren't regressing in what I'd expect to be the common cases and smoothing out those worst cases is a really nice improvement.

@bors r+

**Comment by ssomers** [comment]
With the last commit added on top, rough numbers on my system are almost all down:

bset::tree ttest case|ns/iter before this PR|ns/iter after this PR
--|--|--     
intersect_neg_vs_pos_100|500|60
intersect_neg_vs_pos_10_vs_10k|60|50
intersect_neg_vs_pos_10k|50,000|80
intersect_neg_vs_pos_10k_vs_10|50,000|400
intersect_pos_vs_neg_100|400|60
intersect_pos_vs_neg_10_vs_10k|40,000|50
intersect_pos_vs_neg_10k|40,000|90
intersect_pos_vs_neg_10k_vs_10|60|400
intersect_random_100|700|600
intersect_random_10_vs_10k|40,000|300
intersect_random_10k|200,000|170,000
intersect_random_10k_vs_10|50,000|300
intersect_staggered_100|700|=
intersect_staggered_10_vs_10k|80|170
intersect_staggered_10k|70,000|=
intersect_staggered_10k_vs_10|80|180

As to taking this further with "instead of iterating, find the next element >= the other set's lowest unseen value", no cigar: it slows down intersect_random_10k 2.5 times, and (absolute worst case) intersect_staggered_10k 14x.

The latter could be improved upon by limiting both sets to the range started with the smallest element of the other set. The whole algorithm could work like that: instead of iterating, find the next element >= the other set's lowest unseen value. But then sometimes/often the binary search towards the next element would be worse than iterating, in particular for the staggered test case.

**Comment by ssomers** [comment]
More tweaks:
- expose both intersection strategies as an unstable feature
- benchmark both at the same time, along with the actual strategy chosen
- seriously lower the threshold for the new strategy while still erring on the conservative side
- quantify `are_proportionate_for_intersection` with static assertions instead of unit test

No more ideas at the moment so that's going to be the last batch, unless there are remarks.

**Comment by ssomers** [comment]
Build error seems to be a network timeout, but meanwhile I had another little idea... to be continued

**Pull request closed without merging**